api/types/database.go

```diff
@@ -637,6 +655,55 @@ func (d *DatabaseV3) CheckAndSetDefaults() error {
 	return nil
 }
 
+// checkAndSetDynamoDBDefaults checks DynamoDB configuration and sets defaults.
+func (d *DatabaseV3) checkAndSetDynamoDBDefaults() error {
+	if d.Spec.AWS.AccountID == "" {
+		return trace.BadParameter("database %q AWS account ID is empty", d.GetName())
+	}
+
+	region, err := awsutils.DynamoDBRegionForEndpoint(d.Spec.URI)
+	switch {
+	case err != nil:
+		// when region parsing returns an error but the region is set, it's ok because we can just construct the URI using the region,
+		// so we check if the region is configured to see if this is really a configuration error.
+		if d.Spec.AWS.Region == "" {
+			// the AWS region is empty and we can't derive it from the URI, so this is a config error.
+			return trace.BadParameter("database %q AWS region is empty and cannot be derived from the URI %q",
+				d.GetName(), d.Spec.URI)
+		}
+		if awsutils.IsAWSEndpoint(d.Spec.URI) {
+			// The user configured an AWS URI that which doesn't look like a DynamoDB endpoint.
+			// The URI must look like <service>.<region>.<partition> or <region>.<partition>
+			return trace.Wrap(err)
+		}
+	case d.Spec.AWS.Region == "":
+		// if the AWS region is empty we can just use the region extracted from the URI.
+		d.Spec.AWS.Region = region
+	case d.Spec.AWS.Region != region:
+		// if the AWS region is not empty but doesn't match the URI, this may indicate a user configuration mistake.
+		return trace.BadParameter("database %q AWS region %q does not match the configured URI region %q, "+
+			" omit the URI and it will be derived automatically for the configured AWS region",
+			d.GetName(), d.Spec.AWS.Region, region)
+	}
+
+	if d.Spec.URI == "" || awsutils.IsAWSEndpoint(d.Spec.URI) {
+		// The full URI will be constructed from the region depending on the incoming request,
+		// as: [dynamodb|streams.dynamodb|dax].<region>.<partition>
+		// Therefore we want just the suffix for an AWS endpoint.
+		suffix := awsutils.DynamoDBEndpointSuffixForRegion(d.Spec.AWS.Region)
+		if suffix != d.Spec.URI {
+			logrus.Warnf("overriding database %q AWS URI from %q to %q. "+
```

Comment:
I can remove it. I was not aware that we prefer to avoid logging in `api`, since we do it in `DatabaseV3.CheckAndSetDefaults` for other databases. Why do we prefer to avoid logrus in api?

Reply:
ok hmm. I'd like to log this message. If I remove it now with plans to "eventually" add it back when #12786 is addressed to inject a logger, then it will probably just be forgotten... I don't like adding to the problem, but I'm not sure how else I could log the message without substantial rework

Reply:
Changed the config check/set function so this is obsolete. I do not log anything anymore because I don't override a custom URI.